prow/crier/reporters/gerrit/reporter.go

```diff
@@ -253,6 +265,45 @@ func (c *Client) ShouldReport(ctx context.Context, log *logrus.Entry, pj *v1.Pro
 // Report will send the current prowjob status as a gerrit review
 func (c *Client) Report(ctx context.Context, logger *logrus.Entry, pj *v1.ProwJob) ([]*v1.ProwJob, *reconcile.Result, error) {
 	logger = logger.WithFields(logrus.Fields{"job": pj.Spec.Job, "name": pj.Name})
+
+	// Gerrit reporter hasn't learned how to deduplicate itself from report yet,
+	// will need to block here. Unfortunately need to check after this section
+	// to ensure that the job was not already marked reported by other threads
+	// TODO(chaodaiG): remove postsubmit once
+	// https://github.com/kubernetes/test-infra/issues/22653 is fixed
+	if pj.Spec.Type == v1.PresubmitJob || pj.Spec.Type == v1.PostsubmitJob {
+		key, err := lockKeyForPJ(pj)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get lockkey for job: %w", err)
+		}
+		lock, err := c.prLocks.GetLock(ctx, *key)
+		if err != nil {
+			return nil, nil, err
+		}
+		if err := lock.Acquire(ctx, 1); err != nil {
+			return nil, nil, err
+		}
+		defer lock.Release(1)
+
+		// In the case where several prow jobs from the same PR are finished one
+		// after anther, by the time the lock is acquired, this job might have
+		// already been reported by another worker, refetch this pj to make sure
+		// that no duplicate report is produced
+		pjObjKey := ctrlruntimeclient.ObjectKeyFromObject(pj)
+		if err := c.lister.Get(ctx, pjObjKey, pj); err != nil {
+			if apierrors.IsNotFound(err) {
+				logger.Debug("object no longer exist")
+				return nil, nil, nil
+			}
+
+			return nil, nil, fmt.Errorf("failed to get prowjob %s: %w", pjObjKey.String(), err)
```

Comment:
Why do we return an error in the case where the listing of prowjob failed ("failed to get prowjob"), but we return OK if the prowjob no longer exists (IsNotFound)? I feel like in both cases we both "failed to get a prowjob", so they should return the same error.

Reply:
Job could be GC'ed or deleted for other reasons, this is not a prow error and should not be retried. I've added comments here